ui/src/shared/components/AnnotationEditor.tsx

```diff
@@ -0,0 +1,94 @@
+import React, {PureComponent} from 'react'
+
+import OverlayContainer from 'src/reusable_ui/components/overlays/OverlayContainer'
+import OverlayHeading from 'src/reusable_ui/components/overlays/OverlayHeading'
+import OverlayBody from 'src/reusable_ui/components/overlays/OverlayBody'
+import AnnotationEditorForm from 'src/shared/components/AnnotationEditorForm'
+
+import {Annotation, RemoteDataState} from 'src/types'
+
+interface Props {
+  annotation: Annotation
+  onCancel: () => void
+  onSave: (annotation: Annotation) => Promise<void>
+  onDelete: () => Promise<void>
+}
+
+interface State {
+  draftAnnotation: Annotation | null
+  savingStatus: RemoteDataState
+}
+
+class AnnotationEditor extends PureComponent<Props, State> {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      draftAnnotation: null,
+      savingStatus: RemoteDataState.NotStarted,
+    }
+  }
+
+  public render() {
+    const {annotation, onDelete, onCancel} = this.props
+
+    return (
+      <div className="annotation-editor">
+        <OverlayContainer maxWidth={600}>
+          <OverlayHeading title={'Edit Annotation'}>
+            <div className="annotation-editor--controls">
+              <button className="btn btn-sm btn-default" onClick={onCancel}>
+                Cancel
+              </button>
+              <button
+                className="btn btn-sm btn-success"
+                disabled={!this.canSave || this.isSaving}
+                onClick={this.handleSave}
+                data-test="save-button"
+              >
+                {this.isSaving ? 'Saving...' : 'Save'}
+              </button>
+            </div>
+          </OverlayHeading>
+          <OverlayBody>
+            <AnnotationEditorForm
+              key={annotation.id}
+              annotation={annotation}
+              onSetDraftAnnotation={this.handleSetDraftAnnotation}
+              onDelete={onDelete}
+            />
+          </OverlayBody>
+        </OverlayContainer>
+      </div>
+    )
+  }
+
+  private get canSave(): boolean {
+    return !!this.state.draftAnnotation
+  }
+
+  private get isSaving(): boolean {
+    return this.state.savingStatus === RemoteDataState.Loading
+  }
+
+  private handleSetDraftAnnotation = (
+    draftAnnotation: Annotation | null
+  ): void => {
+    this.setState({draftAnnotation})
+  }
+
+  private handleSave = async (): Promise<void> => {
+    if (!this.canSave) {
+      return
+    }
+
+    const {onSave} = this.props
+    const {draftAnnotation} = this.state
+
+    this.setState({savingStatus: RemoteDataState.Loading})
+
+    await onSave(draftAnnotation)
```

Comment:
hmm the entire component gets removed by its parent after `onSave` resolves, so setting the `savingStatus` would be a set state on unmounted component error. we'd need some way of canceling the whole `handleSave` method on unmount to avoid this.

it does seem a bit funky that the status is never set to done, but I'm going to leave it for now.